src/dbami/db.py

```diff
@@ -406,20 +402,85 @@ async def _update_schema_version(
             await self.execute_sql(schema_sql, conn=conn)
             await self.execute_sql(table_sql, conn=conn)
 
+    async def _update_schema_version(
+        self,
+        version: int,
+        conn: asyncpg.Connection,
+    ) -> None:
+        await self._create_schema_and_version_table(conn)
+
+        version_sql, params = render(
+            "INSERT INTO :version_table (version) VALUES (:version)",
+            version_table=V(self.schema_version_table),
+            version=version,
+        )
+
         await self.execute_sql(version_sql, *params, conn=conn)
 
+    @asynccontextmanager
+    async def migration_lock(
+        self,
+        use_lock: bool = True,
+        timeout_ms: int = 0,
+        **kwargs,
+    ) -> AsyncIterator[asyncpg.Connection]:
+        async with self.get_db_connection(**kwargs) as conn:
+            if not use_lock:
+                yield conn
+                return
+
+            await self._create_schema_and_version_table(conn)
+
+            lock_query, _ = render(
+                """
+                DO $_$
+                    BEGIN
+                        SET lock_timeout = :timeout_ms;
+                        PERFORM pg_advisory_lock(
+                            :dbami_lock_id,
+                            to_regclass(':version_table')::integer
+                        );
+                    END
+                $_$;
+                """,
+                timeout_ms=V(str(timeout_ms)),
+                dbami_lock_id=V(str(DBAMI_LOCK_ID)),
+                version_table=V(self.schema_version_table),
+            )
+            unlock_query, unlock_params = render(
+                """
+                SELECT pg_advisory_unlock(
+                    :dbami_lock_id,
+                    to_regclass(:version_table)::oid::integer
+                )
+                """,
+                dbami_lock_id=DBAMI_LOCK_ID,
+                version_table=self.schema_version_table,
+            )
+
+            try:
+                await self.execute_sql(lock_query, conn=conn)
+            except asyncpg.exceptions.LockNotAvailableError:
+                raise exceptions.LockError(
+                    "Unable to acquire a migration lock because it is held by another "
+                    "user."
+                )
+            yield conn
+            await self.execute_sql(unlock_query, *unlock_params, conn=conn)
+
     async def migrate(
         self,
         target: Optional[int] = None,
         direction: Union[Literal["up"], Literal["down"], None] = None,
+        use_lock: bool = True,
```

Comment:
Add timeout argument explicitly.

Reply:
Done. As discussed, also updated the `up` and `rollback` CLI commands to include the `no_lock` flag and the `timout_ms` option since they both call `db.migrate`.